### ISSUE TITLE
Semaphore usage in SttpOauth2ClientCredentialsCatsBackend

### ISSUE THREAD

**Issue by matwojcik** [ocadotechnology/sttp-oauth2#121]
I'd like to open a discussion about the usage of `Semaphore` in `SttpOauth2ClientCredentialsCatsBackend`.

The aim of this was to prevent parallel calls for token e.g. at application startup or when the token expired. When there is nothing in cache and multiple threads are calling this, without a `Semaphore` all of them would unnecessarily do a http call for token in paralel.

However token life is usually quite long, let's say half an hour. Then all calls during this half an hour are blocked from paralel access to read only resource in memory. This way for a sake of app startup and then one moment every half an hour, we pay a cost for half an hour of obtaining valid token from cache.

I'm not 100% convinced this is a right choice. Of course current cache implementation is in memory and is quite quick, so we don't block for long, but if someone used some more sophisticated caching mechanism, they would unnecessarily pay a cost there.

Some middle-way solution would be creating `Semaphore` with `n` e.g. 4 - that would prevent from flooding token service in worst cases, but would enable parallel access in regular situations.

What would be a good thing I guess is doing some performance test, that would measure what is better performance-wise for an application.

**Comment by kubukoz** [comment]
Getting the token from memory is as fast as possible with a concurrent resource (it's a Ref so backed by an AtomicReference) that I think for the current implementation there's no need to complicate the design.

If we want to support all kinds of caches, sure, something should be done. I have something different in mind though: changing what's guarded by the semaphore.

Currently every call has this:

1. enter semaphore
2. cache.get
3. check if expired (assuming true)
4. fetch token
5. update cache
6. exit semaphore

I think we could do this:

1. cache.get
2. check if expired (assuming true)
3. enter semaphore
4. cache.get (ensure nobody else has refreshed the token)
5. fetch token
6. update cache
7. exit semaphore

There's no need for an `n` greater than 1, although we have the extra cost of checking the cache N times for N concurrent attempts to refresh the token. With a TTL of 30 minutes this seems to be a no-brainer.

**Comment by matwojcik** [comment]
Yeah, I think it's good approach, I thought about changing the guard, but without this additional check after getting the permission it was pointless.

Internally we're playing with caching whole provider instead, after it gets proven on production we'll probably introduce it here.
